newm-server/src/main/kotlin/io/newm/server/features/cardano/parser/NFTSongParser.kt

```diff
@@ -0,0 +1,142 @@
+            }
+
+            "release" -> {
+                // only available in CIP60-V2 or above
+                item.childrenList.forEach(parseSongItem)
+            }
+
+            else -> parseSongItem(item)
+        }
+    }
+    return files.mapNotNull { file ->
+        val title = file.songTitle ?: songTitle ?: file.name
+        if (title == null || file.src == null || image == null) {
+            logger.warn { "Skipped invalid PolicyID: ${asset.policy}, assetName: ${asset.name}, title = $title, src: ${file.src}, image = $image" }
+            null
+        } else {
+            NFTSong(
+                id = UUID.nameUUIDFromBytes(file.src.toByteArray()),
+                policyId = asset.policy,
+                assetName = asset.name,
+                amount = asset.amount.toLong(),
+                title = title,
+                audioUrl = file.src.toResourceUrl(),
+                imageUrl = image.toResourceUrl(),
+                duration = (file.songDuration ?: songDuration)?.let { Duration.parse(it).inWholeSeconds } ?: 0,
+                artists = file.artists ?: artists.toList(),
+                genres = file.genres ?: genres.toList(),
+                moods = file.mood?.let { listOf(it) } ?: moods.toList()
+            )
+        }
+    }
+}
+
+private operator fun List<LedgerAssetMetadataItem>.get(key: String): LedgerAssetMetadataItem? =
+    firstOrNull { it.key == key }
+
+private fun List<LedgerAssetMetadataItem>.value(key: String): String? = this[key]?.value
+
+private fun LedgerAssetMetadataItem.values(): List<String> = childrenList?.map { it.value }.orEmpty()
+
+private fun LedgerAssetMetadataItem.artists(): List<String> =
+    childrenList?.mapNotNull { it.childrenList.value("name") }.orEmpty()
+        .ifEmpty { values() }
+        .ifEmpty { value?.let(::listOf) }.orEmpty()
+
+private fun LedgerAssetMetadataItem.audioFiles(): List<File> =
+    childrenList?.filter { it.isAudioFile }?.map { it.childrenList.toFile() }.orEmpty()
+
+private val LedgerAssetMetadataItem.isAudioFile: Boolean
+    get() = childrenList.value("mediaType")?.startsWith("audio/") == true
+
+private data class File(
+    val name: String?,
+    val src: String?,
+    var songTitle: String?,
+    var songDuration: String?,
+    val artists: List<String>?,
+    val genres: List<String>?,
+    val mood: String?
+)
+
+private fun List<LedgerAssetMetadataItem>.toFile(): File {
+    // "song" only available in CIP60-V2 or above
+    val song = this["song"]?.childrenList ?: this
+    return File(
+        name = value("name"),
+        src = value("src"),
+        songTitle = song.value("song_title"),
+        songDuration = song.value("song_duration"),
+        artists = song["artists"]?.artists(),
+        genres = song["genres"]?.values(),
+        mood = song.value("mood")
+    )
+}
+
+private fun String.toResourceUrl(): String = when {
+    startsWith("ipfs://") -> "https://bcsh.mypinata.cloud/ipfs/${substring(7)}"
```

Comment:
`https://ipfs.io/ipfs/${substring(7)}`

Do not publish my ipfs gateway to github please.

Reply:
Foxed